lib/web/apiserver.go

```diff
@@ -1013,40 +1011,47 @@ func (h *Handler) oidcLoginConsole(w http.ResponseWriter, r *http.Request, p htt
 }
 
 func (h *Handler) oidcCallback(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
-	logger := newPackageLogger("oidc")
+	logger := h.log.WithField("auth", "oidc")
 	logger.Debug("Callback start.")
 
 	response, err := h.cfg.ProxyClient.ValidateOIDCAuthCallback(r.URL.Query())
 	if err != nil {
-		logger.WithError(err).Warn("Error while processing callback.")
-
-		message := "Unable to process callback from OIDC provider."
-		// redirect to an error page
-		pathToError := url.URL{
-			Path:     "/web/msg/error/login_failed",
-			RawQuery: url.Values{"details": []string{message}}.Encode(),
-		}
-		http.Redirect(w, r, pathToError.String(), http.StatusFound)
+		logger.WithError(err).Error("Error while processing callback.")
+		http.Redirect(w, r, loginFailedBadCallbackURL, http.StatusFound)
 		return nil, nil
 	}
+
 	// if we created web session, set session cookie and redirect to original url
 	if response.Req.CreateWebSession {
-		err = csrf.VerifyToken(response.Req.CSRFToken, r)
-		if err != nil {
-			logger.WithError(err).Warn("Unable to verify CSRF token.")
-			return nil, trace.AccessDenied("access denied")
+		logger.Info("Redirecting to web browser.")
+
+		if err := csrf.VerifyToken(response.Req.CSRFToken, r); err != nil {
+			logger.WithError(err).Error("Unable to verify CSRF token.")
+			http.Redirect(w, r, loginFailedURL, http.StatusFound)
+			return nil, nil
 		}
 
-		logger.Info("Callback redirecting to web browser.")
 		if err := SetSessionCookie(w, response.Username, response.Session.GetName()); err != nil {
-			return nil, trace.Wrap(err)
+			logger.WithError(err).Error("Unable to set session cookie.")
+			http.Redirect(w, r, loginFailedURL, http.StatusFound)
+			return nil, nil
 		}
-		return nil, httplib.SafeRedirect(w, r, response.Req.ClientRedirectURL)
+
+		if err := httplib.SafeRedirect(w, r, response.Req.ClientRedirectURL); err != nil {
+			logger.WithError(err).Error("Error parsing redirect URL.")
+			http.Redirect(w, r, loginFailedURL, http.StatusFound)
+		}
```

Comment:
Nit: I know this is the last "if" block of the function, but I would add explicit `return nil, nil` inside the block anyway, in case someone adds more code below and forgets to add return here.

Reply:
revised